---
title: Intratumoral T cell and DC composition associated with ICB responsiveness in breast cancer
date: "`r Sys.Date()`"
author: "M. Andreatta and S. Carmona"
output:
  rmdformats::readthedown:
    self-contained: true
    highlight: haddock
    thumbnails: false
    css: styles.css
knit: (function(input_file, encoding) {
  out_dir <- 'docs';
  rmarkdown::render(input_file, encoding=encoding,
  output_file=file.path(dirname(input_file), out_dir, 'Bassez_BC.html'))})
---

```{r setup, echo=FALSE}
#install.packages("rmdformats")
#Template markdown setup
library(knitr)
library(rmdformats)
library(formatR)

## Global options
options(max.print="75")
opts_chunk$set(echo=TRUE,
	             cache=TRUE,
               cache.lazy=FALSE,
               prompt=FALSE,
               tidy=TRUE,
               comment=NA,
               message=FALSE,
               warning=FALSE,
               dev='png')
opts_knit$set(width=75)

```

# Background

The single-cell dataset generated by [Bassez et al. (2021) Nature Medicine](https://www.nature.com/articles/s41591-021-01323-8) contains single-cell data from breast carcinoma biopsies (whole tumors) from 29 patients taken immediately before (pre-treatment) and 9 +/- 2 days after (on-treatment) **receiving anti-PD1 immunotherapy**. Lacking an objective measure of clinical benefit following therapy, in this study the authors use T cell expansion following treatment as a proxy for response.

[Full Data Original Source](https://biokey.lambrechtslab.org/en/data-access)

[Case Study Dataset Internal link](https://drive.switch.ch/index.php/s/7jdqnvPZuJrriZB)

# Goal
To compare the T cell and DC subtype composition between responders and non-responders using `ProjecTILs` and reference maps for [CD8 T cells](https://doi.org/10.6084/m9.figshare.21931875), [CD4 T cells](https://doi.org/10.6084/m9.figshare.21981536) and [Dendritic Cells](https://doi.org/10.6084/m9.figshare.22040801.v1)

# R Environment

```{r, message=FALSE, warning=F, results=FALSE}
library(renv)
renv::restore()

library(remotes)
library(patchwork)
library(ggplot2)
library(reshape2)
library(Seurat)
library(ProjecTILs)
library(parallel)

#remotes::install_github("carmonalab/ProjecTILs",ref="9e55c097d191deb5c75bbb316b7495621e205c08")
```

# scRNA-seq data preparation

```{r}
ddir <- "input/Bassez_breast_data"

#Count matrices
f1 <- sprintf("%s/1863-counts_cells_cohort1.rds", ddir)

cohort1 <- readRDS(f1)
dim(cohort1)

#Metadata
meta1 <- read.csv(sprintf("%s/1872-BIOKEY_metaData_cohort1_web.csv", ddir))
rownames(meta1) <- meta1$Cell

head(meta1)

data.seurat <- CreateSeuratObject(cohort1, project="Cohort1_IT", meta.data = meta1)
data.seurat <- NormalizeData(data.seurat)

```


Subset pre-treatment biopsies
```{r}
table(data.seurat$timepoint)
data.seurat <- subset(data.seurat, subset=timepoint=="Pre")
```

```{r}
table(data.seurat$cellType)
```


Optional -  remove samples that are too small and downsample large ones, in order to have similar contribution from all patients/samples. Downsampling large samples also speeds up downstream computations.

```{r, eval=F}
ds <- 1000
min.cells <- 200

tab <- table(data.seurat$patient_id)
keep <- names(tab)[tab > min.cells]
data.seurat <- subset(data.seurat, patient_id %in% keep)


Idents(data.seurat) <- "patient_id"
data.seurat <- subset(data.seurat, cells = WhichCells(data.seurat,downsample = ds))
table(data.seurat$patient_id)
```

# ProjecTILs analysis

## Load reference maps

Here we will project the query data onto human reference TIL atlas for CD4 and CD8 T cells, with the goal to interpret T cell diversity in the context of reference T cell subtypes. First, load the reference maps:
```{r fig.height=4, fig.width=9}
options(timeout = max(900, getOption("timeout")))

#download.file("https://figshare.com/ndownloader/files/38921366", destfile = "CD8T_human_ref_v1.rds")
ref.cd8 <- load.reference.map("CD8T_human_ref_v1.rds")

#download.file("https://figshare.com/ndownloader/files/39012395", destfile = "CD4T_human_ref_v1.rds")
ref.cd4 <- load.reference.map("CD4T_human_ref_v1.rds")

#download.file("https://figshare.com/ndownloader/files/39120752", destfile = "DC_human_ref_v1_light.rds")
ref.dc <- load.reference.map("DC_human_ref_v1_light.rds")

a <- DimPlot(ref.cd8, cols=ref.cd8@misc$atlas.palette, label = T) +
  theme(aspect.ratio = 1) + ggtitle("CD8 T reference") + NoLegend()

b <- DimPlot(ref.cd4, cols=ref.cd4@misc$atlas.palette, label = T) +
  theme(aspect.ratio = 1) + ggtitle("CD4 T reference") + NoLegend()

c <- DimPlot(ref.dc, cols=ref.dc@misc$atlas.palette, label = T) +
  theme(aspect.ratio = 1) + ggtitle("DC reference") + NoLegend()

a | b | c
```

Show key marker genes for the reference subtypes
```{r fig.height=10, fig.width=6.5}
DefaultAssay(ref.cd8) <- "RNA"

markers.cd8 <- c("SELL","TCF7","LEF1","CCR7","S1PR1","LMNA","IL7R","GZMK",
           "FGFBP2",'FCGR3A','XCL1',"XCL2","CD200","CRTAM","GNG4","TOX",'PDCD1',"HAVCR2",
           "GZMB","PRF1","LAG3",'KLRB1','TRAV1-2')

a <- VlnPlot(ref.cd8, features = markers.cd8, stack = TRUE, flip = TRUE,
             fill.by = "ident", cols = ref.cd8@misc$atlas.palette) +
             NoLegend()
DefaultAssay(ref.cd8) <- "integrated"

DefaultAssay(ref.cd4) <- "RNA"
Idents(ref.cd4) <- "functional.cluster"

markers.cd4 <- c("SELL","TCF7","S1PR1","KLF2","IL7R","CCL5","GZMK","EOMES",
           "TBX21","CXCR6","GZMA","GZMB","PRF1","GNLY",
           "CXCL13","PDCD1","TOX","LAG3","HAVCR2","KLRB1","IL17A","FOXP3","IL2RA")

b <- VlnPlot(ref.cd4, features = markers.cd4, stack = TRUE,flip = TRUE,
             fill.by = "ident", cols = ref.cd4@misc$atlas.palette) +
             NoLegend()
DefaultAssay(ref.cd4) <- "integrated"


DefaultAssay(ref.dc) <- "RNA"
Idents(ref.dc) <- "functional.cluster"

markers.dc <- c("CLEC9A","XCR1","CADM1","CLEC10A","CD1C","CD1A",
                 "LAMP3","CCR7","FSCN1","S100A8","S100A9","VCAN","FCN1",
                 "GZMB","TSPAN13","TCF4","PPP1R14A","AXL")

markers.dc.2 <- c("CLEC9A","XCR1","CADM1","CLEC10A","FCGR2A","FCER1A","CD1C","CD1A","CD207",
                 "LAMP3","CCR7","FSCN1","GZMB","LILRA4","TCF4","PPP1R14A","AXL",
           "S100A8","S100A9","VCAN","FCN1","ITGAX","HLA-DRA", "HLA-DQA1","HLA-DQB1")

c <- VlnPlot(ref.dc, features = markers.dc, stack = TRUE,flip = TRUE,
             fill.by = "ident", cols = ref.dc@misc$atlas.palette) +
             NoLegend()

DefaultAssay(ref.dc) <- "integrated"

a | b | c
```

## Reference-based analysis

To classify CD4 and CD8 T cell subtypes, we will project CD4 and CD8 T cells separately into the two reference maps. Running `ProjecTILs.classifier` sequentially on the two maps allows combining the subtype predictions for a complete annotation of T cells subtypes.

For optimal batch-effect correction, we recommend projecting each patient/batch separately (set `split.by="patient_id"`). For large samples, `Run.ProjecTILs` can take several minutes. Set the number of parallel jobs with *ncores* according to your computational resources.


```{r}
data.seurat.list <- SplitObject(data.seurat, split.by = "patient_id")

```

```{r}
ncores=8

#Classify CD8 T subtypes

data.seurat.list <- mclapply(data.seurat.list,
     FUN = function(x) {
        ProjecTILs.classifier(x, ref.cd8, ncores=1)
     }, mc.cores = ncores)
  
```

```{r, eval=F}
#Classify CD8 T subtypes
data.seurat <-  ProjecTILs.classifier(data.seurat, ref.cd8,
                                  ncores=ncores, split.by = "patient_id")
```

After running the classifier with the CD8T reference, we obtain CD8 T cell subtype annotations:
```{r}
table(data.seurat.list[[1]]$functional.cluster, useNA="ifany")
```

We can now run the classifier on the same object with the CD4 T reference, specifying the parameter `overwrite=FALSE`. This adds labels for the CD4 T cells without replacing the CD8 T cell subtypes. Note that only cells that can be unequivocally assigned to one cell type are labeled; the remaining cells (including those that receive a label by more than one reference) are assigned to NA.


```{r}
data.seurat.list <- mclapply(data.seurat.list,
     FUN = function(x) {
        ProjecTILs.classifier(x, ref.cd4,
                                  ncores=1, 
                                  overwrite = FALSE)
     }, mc.cores = ncores)
  
```


The labels now contain both CD4T and CD8T subtypes:
```{r}
table(data.seurat.list[[1]]$functional.cluster, useNA="ifany")
```


We can now run the classifier on the same object for the DC reference


```{r}
data.seurat.list <- mclapply(data.seurat.list,
     FUN = function(x) {
        ProjecTILs.classifier(x, ref.dc,
                                  ncores=1, 
                                  overwrite = FALSE)
     }, mc.cores = ncores)
  
```




```{r}
data.seurat <- merge(data.seurat.list[[1]],data.seurat.list[2:length(data.seurat.list)])
```


The labels now contain CD4T, CD8T, and DC subtypes:
```{r}
table(data.seurat$functional.cluster, useNA="ifany")
```




We can verify the expression profile using a panel of marker genes

This one is tailored for T cells
```{r fig.height=9, fig.width=13, eval-F}
genes4radar <- c("CD4","CD8A","TCF7","CCR7","IL7R","LMNA","GZMA","GZMK","FGFBP2","XCL1",
                 "CRTAM","TOX","PDCD1","HAVCR2","PRF1","GLNY","GZMB","TRAV1-2","KLRB1","FOXP3")

plot.states.radar(ref=ref.cd8, query=data.seurat, genes4radar = genes4radar)
plot.states.radar(ref=ref.cd4, query=data.seurat, genes4radar = genes4radar)
```

This one is a broader panel of marker genes for T cells and DCs
```{r fig.height=9, fig.width=13}
genes4radar <- c("CD4","CD8A","TCF7","CCR7","IL7R","GZMA","GZMK","GZMB","PRF1","GLNY","XCL1", "CRTAM","TOX","PDCD1","HAVCR2","TRAV1-2","KLRB1","FOXP3","CLEC10A","CD1A","LAMP3","S100A9","AXL","HLA-DRA")

plot.states.radar(ref=ref.cd8, query=data.seurat, genes4radar = genes4radar)
plot.states.radar(ref=ref.cd4, query=data.seurat, genes4radar = genes4radar)
plot.states.radar(ref=ref.dc, query=data.seurat, genes4radar = genes4radar)
```


Compare subtype composition of responders vs. non-responders (based on clonal expansion) on pre-therapy samples
```{r fig.height=4, fig.width=6.5}
a <- ref.cd8@misc$atlas.palette
b <- ref.cd4@misc$atlas.palette
c <- ref.dc@misc$atlas.palette
cols <- c(a,b,c)

by.exp <- SplitObject(data.seurat, split.by = "expansion")

tb <-  table(factor(by.exp$E$functional.cluster, levels = names(cols)))
tb <- tb * 100/sum(tb)
tb.m <- reshape2::melt(tb)
colnames(tb.m) <- c("Cell_state", "Perc_cells")

p1 <- ggplot(tb.m, aes(x = Cell_state, y = Perc_cells, 
            fill = Cell_state)) + geom_bar(stat = "identity") + 
            theme_bw() + scale_fill_manual(values = cols) +
            theme(axis.text.x = element_blank(), legend.position = "none") +
            ggtitle("Responders") + ylim(0,50)


tb <-  table(factor(by.exp$NE$functional.cluster, levels = names(cols)))
tb <- tb * 100/sum(tb)
tb.m <- reshape2::melt(tb)
colnames(tb.m) <- c("Cell_state", "Perc_cells")

p2 <- ggplot(tb.m, aes(x = Cell_state, y = Perc_cells, 
            fill = Cell_state)) + geom_bar(stat = "identity") + 
            theme_bw() + scale_fill_manual(values = cols) +
            theme(axis.text.x = element_blank(), legend.position = "right") +
            ggtitle("Non-Responders") + ylim(0,50)

pp <- p1 | p2

ggsave("plots/Bassez2021_barplots_RNR.pdf",width = 20, height = 10)

pp <- (p1 | p2) & coord_cartesian(ylim=c(0, 2))

pp

ggsave("plots/Bassez2021_barplots_RNR_zoom5.pdf",width = 20, height = 10)
```

A more compact way of displaying enrichment/depletion of specific subtypes is to calculate and visualize **fold-changes** of TIL composition in **responders vs non-responders**.
However, note that some frequencies are very small and might not differ significantly; more proper statistical analyses for differential cell type abundance are recommended.

```{r fig.height=3.5, fig.width=5}
which.types <- table(data.seurat$functional.cluster) > 20  # disregard small populations

states_all <- names(cols)

cols_use <- cols[names(which.types)][which.types]

norm.c <- table(by.exp$NE$functional.cluster)/sum(table(by.exp$NE$functional.cluster))
norm.q <- table(by.exp$E$functional.cluster)/sum(table(by.exp$E$functional.cluster))

foldchange <- norm.q[which.types]/norm.c[which.types]
foldchange <- sort(foldchange, decreasing = T)

tb.m <- reshape2::melt(foldchange)
colnames(tb.m) <- c("Cell_state", "Fold_change")
pll <- list()
gg <- ggplot(tb.m, aes(x = Cell_state, y = Fold_change, fill = Cell_state)) +
  geom_bar(stat = "identity") +
    scale_fill_manual(values = cols_use) + theme_bw() +
  geom_hline(yintercept = 1) + scale_y_continuous(trans = "log2") +
    theme(axis.text.x = element_blank(), legend.position = "left") +
  ggtitle("Responders vs. Non-responders")

gg

ggsave("plots/Bassez2021_barplots_RNR_foldchange.pdf",width = 10, height = 10)
```

The analysis suggests that ICB-responding tumors have a higher proportion of CD8 exhausted (CD8.TEX), CD8 precursor exhausted (CD8.TPEX), CD4 cytotoxic exhausted (CD4.CTL_Exh) T cells and a subtype of dendritic cells DC3 (characterized by expresssion of LAMP3, CCR7, IL7R); on the other hand, CD8 TEMRA, CD4 and CD8 Naive-like, DC2s (both CLEC10A and CD1A) and MonoDCs seem to be more abundant in non-responders.


# Exploratory subtype composition using Correspondence Analysis

Do patients cluster based on their subtype composition? are subtype composition "classes" related to response to ICB? we can perform correspondence analysis (CA) to try to answer these questions, and group patients based on their composition of T cell and DC subtypes.

```{r fig.width=12, fig.height=7}
library("FactoMineR")
min.cells <- 30

useNotAnnotatedCells = T
normalizeCellCounts <- F

#Only use samples with sufficient amount of cells
tab <- table(data.seurat$patient_id)
samples.use <- names(tab)[tab>min.cells]
data.seurat <- subset(data.seurat, subset = patient_id %in% samples.use)

if(useNotAnnotatedCells){
  if(normalizeCellCounts){
    subtype_comp_table <- prop.table(table(data.seurat$functional.cluster,
                                       data.seurat$patient_id,useNA="ifany"), margin = 2)
  } else {
    subtype_comp_table <- table(data.seurat$functional.cluster,data.seurat$patient_id,useNA="ifany")  
  }
  rownames(subtype_comp_table)[length(rownames(subtype_comp_table))] <- "Others"
} else {
   if(normalizeCellCounts){
    subtype_comp_table <- prop.table(table(data.seurat$functional.cluster,
                                       data.seurat$patient_id), margin = 2)
   } else {
    subtype_comp_table <- table(data.seurat$functional.cluster,data.seurat$patient_id)  
  }
}


head(t(subtype_comp_table))

res.CA <- CA(t(subtype_comp_table), graph = FALSE)
options("max.print"=150)
res.CA$col$coord

a <- plot(res.CA) + theme(aspect.ratio = 1) 

#Color by response
coords <- as.data.frame(res.CA$row$coord)
coords.ct <- as.data.frame(res.CA$col$coord)
coords.all <- rbind(coords[,1:5], coords.ct[,1:5])

is.R <- unique(data.seurat$patient_id[data.seurat$expansion=='E'])
is.NR <- unique(data.seurat$patient_id[data.seurat$expansion=='NE'])
coords[['Responder']] <- "n/a"
coords$Responder[rownames(coords) %in% is.R] <- "Responder"
coords$Responder[rownames(coords) %in% is.NR] <- "Non-responder"
coords[['Patient']] <- rownames(coords)

b <- ggplot(coords, aes(x = `Dim 1`, y = `Dim 2`, color = Responder)) +
  geom_point(size=4) + theme_bw() + theme(aspect.ratio = 1) +
  scale_color_manual(values=c("#E9E9E9","#fb0505","#08e579")) +
  xlim(min(coords.all[,'Dim 1']), max(coords.all[,'Dim 1'])) +
  ylim(min(coords.all[,'Dim 2']), max(coords.all[,'Dim 2']))

a | b
ggsave("plots/Bassez2021_CA.pdf",width = 20, height = 10)
ggsave("plots/Bassez2021_CA.png",width = 20, height = 10)

```

This analysis suggests some interesting statistical associations such as a pattern enriched in CD8 TEX, CD8 TPEX, CD4 Treg and DC3,  and another pattern enriched in CD8 and CD4 Naive states and CD2.
The comparison of absolute frequencies (i.e. including "other"/unclassified cells) vs. relative frequencies (over considered populations, here T cells and DCs) suggests that total immune infiltrate does not have a strong association with response. Rather, the relative composition of T cells and DC subtypes leads to a larger separation of R vs NR.


# Exploratory subtype composition using Correspondence Analysis - per cell type


```{r fig.width=12, fig.height=7}

min.cells <- 30

useNotAnnotatedCells = T
normalizeCellCounts <- F

#Only use samples with sufficient amount of cells
tab <- table(data.seurat$patient_id)
samples.use <- names(tab)[tab>min.cells]
data.seurat <- subset(data.seurat, subset = patient_id %in% samples.use)

useMap <- "ref.cd4"
use.labels <- c(levels(get(useMap)$functional.cluster),NA)

if(useNotAnnotatedCells){
  if(normalizeCellCounts){
    subtype_comp_table <- prop.table(table(data.seurat$functional.cluster[which(data.seurat$functional.cluster %in% use.labels)],
                                       data.seurat$patient_id[which(data.seurat$functional.cluster %in% use.labels)],useNA="ifany"), margin = 2)
  } else {
    subtype_comp_table <- table(data.seurat$functional.cluster[which(data.seurat$functional.cluster %in% use.labels)],data.seurat$patient_id[which(data.seurat$functional.cluster %in% use.labels)],useNA="ifany")  
  }
  rownames(subtype_comp_table)[length(rownames(subtype_comp_table))] <- "Others"
} else {
   if(normalizeCellCounts){
    subtype_comp_table <- prop.table(table(data.seurat$functional.cluster[data.seurat$functional.cluster %in% use.labels],
                                       data.seurat$patient_id[which(data.seurat$functional.cluster %in% use.labels)]), margin = 2)
   } else {
    subtype_comp_table <- table(data.seurat$functional.cluster[data.seurat$functional.cluster %in% use.labels],data.seurat$patient_id[which(data.seurat$functional.cluster %in% use.labels)])  
  }
}


head(t(subtype_comp_table))

res.CA <- CA(t(subtype_comp_table), graph = FALSE)
options("max.print"=150)
res.CA$col$coord

a <- plot(res.CA, cex=1, autoLab = "no") + theme(aspect.ratio = 1)
#a <- plot(res.CA, cex=1, autoLab = "yes", max.overlaps = Inf) + theme(aspect.ratio = 1) 

#Color by response
coords <- as.data.frame(res.CA$row$coord)
coords.ct <- as.data.frame(res.CA$col$coord)
coords.all <- rbind(coords[,1:5], coords.ct[,1:5])

is.R <- unique(data.seurat$patient_id[data.seurat$expansion=='E'])
is.NR <- unique(data.seurat$patient_id[data.seurat$expansion=='NE'])
coords[['Responder']] <- "n/a"
coords$Responder[rownames(coords) %in% is.R] <- "Responder"
coords$Responder[rownames(coords) %in% is.NR] <- "Non-responder"
coords[['Patient']] <- rownames(coords)

b <- ggplot(coords, aes(x = `Dim 1`, y = `Dim 2`, color = Responder)) +
  geom_point(size=4) + theme_bw() + theme(aspect.ratio = 1) +
  scale_color_manual(values=c("#E9E9E9","#fb0505","#08e579")) +
  xlim(min(coords.all[,'Dim 1']), max(coords.all[,'Dim 1'])) +
  ylim(min(coords.all[,'Dim 2']), max(coords.all[,'Dim 2']))

a | b
ggsave(paste("plots/Bassez2021_CA",useMap,useNotAnnotatedCells,normalizeCellCounts,".pdf",sep="_"),width = 20, height = 10)
ggsave(paste("plots/Bassez2021_CA",useMap,useNotAnnotatedCells,normalizeCellCounts,".png",sep="_"),width = 20, height = 10)

```

Looking at each cell type at a time, it seems that both CD8 and CD4 T cell frequencies separate to some extent responders, whereas DC subtypes freq. alone does not.





# Discussion

Our automated ProjecTILs analysis of this patient cohort indicate that compared to non-responders, early responders to PD-1 blockade (in terms of T cell expansion) have higher relative amounts of exhausted T cells (both CD8+ and CD4+) infiltrating their tumors, in agreement with previous studies (Daud et al. 2016; Thommen et al. 2018; Kumagai et al. 2020; Bassez et al. 2020).  This supports the notion that patients with a pre-existing tumor-specific T cell response in their tumors are more likely to respond to PD-1 blockade therapy. 

Because tumoricidal CD8 Tex are derived from Tpex (Siddiqui et al. 2019; Miller et al. 2019), the presence of CD8 Tex seems to be a good proxy for the presence of smaller populations of quiescent Tpex, that have the ability to proliferate and differentiate into Tex cells following PD-1 blockade (Siddiqui et al. 2019; Miller et al. 2019). Indeed we observed a small population of Tpex with increased levels R vs RN. Because there are very few Tpex their gene profile might be less robust than others, but compared to Tex they display higher levels of Tpex-associated genes, such as *TCF7* and *XCL1*, and lower levels of terminal-exhaustion markers such as *HAVCR2* and *ENTPD1*.

# Further reading

Dataset original publication -  [Bassez et al](https://www.nature.com/articles/s41591-021-01323-8)

ProjecTILs case studies - [INDEX](https://carmonalab.github.io/ProjecTILs_CaseStudies/) - [Repository](https://github.com/carmonalab/ProjecTILs_CaseStudies)

The ProjecTILs method [Andreatta et. al (2021) Nat. Comm.](https://www.nature.com/articles/s41467-021-23324-4) and [code](https://github.com/carmonalab/ProjecTILs)


# References

* Bassez, A., Vos, H., Van Dyck, L. et al. A single-cell map of intratumoral changes during anti-PD1 treatment of patients with breast cancer. Nat Med 27, 820–832 (2021). https://doi.org/10.1038/s41591-021-01323-8

* Daud, A.I., Loo, K., Pauli, M.L., Sanchez-Rodriguez, R., Sandoval, P.M., Taravati, K., Tsai, K., Nosrati, A., Nardo, L., Alvarado, M.D., Algazi, A.P., Pampaloni, M.H., Lobach, I.V., Hwang, J., Pierce, R.H., Gratz, I.K., Krummel, M.F., Rosenblum, M.D., 2016. Tumor immune profiling predicts response to anti-PD-1 therapy in human melanoma. J. Clin. Invest. 126, 3447–3452. https://doi.org/10.1172/JCI87324

* Gros, A., Robbins, P.F., Yao, X., Li, Y.F., Turcotte, S., Tran, E., Wunderlich, J.R., Mixon, A., Farid, S., Dudley, M.E., Hanada, K., Almeida, J.R., Darko, S., Douek, D.C., Yang, J.C., Rosenberg, S. a, 2014. PD-1 identifies the patient-specific in filtrating human tumors. J. Clin. Invest. 124, 2246–59. https://doi.org/10.1172/JCI73639.2246

* Miller, B.C., Sen, D.R., Al Abosy, R., Bi, K., Virkud, Y.V., LaFleur, M.W., Yates, K.B., Lako, A., Felt, K., Naik, G.S., Manos, M., Gjini, E., Kuchroo, J.R., Ishizuka, J.J., Collier, J.L., Griffin, G.K., Maleri, S., Comstock, D.E., Weiss, S.A., Brown, F.D., Panda, A., Zimmer, M.D., Manguso, R.T., Hodi, F.S., Rodig, S.J., Sharpe, A.H., Haining, W.N., 2019. Subsets of exhausted CD8+ T cells differentially mediate tumor control and respond to checkpoint blockade. Nat. Immunol. 20, 326–336. https://doi.org/10.1038/s41590-019-0312-6

* Siddiqui, I., Schaeuble, K., Chennupati, V., Fuertes Marraco, S.A., Calderon-Copete, S., Pais Ferreira, D., Carmona, S.J., Scarpellino, L., Gfeller, D., Pradervand, S., Luther, S.A., Speiser, D.E., Held, W., 2019. Intratumoral Tcf1+PD-1+CD8+ T Cells with Stem-like Properties Promote Tumor Control in Response to Vaccination and Checkpoint Blockade Immunotherapy. Immunity 50, 195–211.e10. https://doi.org/10.1016/J.IMMUNI.2018.12.021

* Thommen, D.S., Koelzer, V.H., Herzig, P., Roller, A., Trefny, M., Dimeloe, S., Kiialainen, A., Hanhart, J., Schill, C., Hess, C., Prince, S.S., Wiese, M., Lardinois, D., Ho, P.C., Klein, C., Karanikas, V., Mertz, K.D., Schumacher, T.N., Zippelius, A., 2018. A transcriptionally and functionally distinct pd-1 + cd8 + t cell pool with predictive potential in non-small-cell lung cancer treated with pd-1 blockade. Nat. Med. 24, 994. https://doi.org/10.1038/s41591-018-0057-z

* Kumagai, S., Togashi, Y., Kamada, T. et al. The PD-1 expression balance between effector and regulatory T cells predicts the clinical efficacy of PD-1 blockade therapies. Nat Immunol 21, 1346–1358 (2020). https://doi.org/10.1038/s41590-020-0769-3
